modules/apps/commerce/commerce-shipping-engine/src/mail/java/com/liferay/commerce/shipping/engine/internal/FunctionCommerceShippingEngine.java

```diff
@@ -0,0 +1,335 @@
+			JSONObject jsonObject = _jsonFactory.createJSONObject(
+				new String(
+					_portalCatapult.launch(
+						currentUser.getCompanyId(), Http.Method.POST,
+						_functionCommerceShippingEngineConfiguration.
+							oAuth2ApplicationExternalReferenceCode(),
+						JSONUtil.put("locale", locale),
+						_functionCommerceShippingEngineConfiguration.
+							shippingEngineLocalizedNamePath(),
+						currentUser.getUserId()
+					).get()));
+
+			return jsonObject.getString("name");
+		}
+		catch (Exception exception) {
+			_log.error(exception);
+		}
+
+		return StringPool.BLANK;
+	}
+
+	@Activate
+	protected void activate(Map<String, Object> properties) {
+		_functionCommerceShippingEngineConfiguration =
+			ConfigurableUtil.createConfigurable(
+				FunctionCommerceShippingEngineConfiguration.class, properties);
+	}
+
+	@Deactivate
+	protected void deactivate() throws PortalException {
+		String key = getKey();
+
+		if (key == null) {
+			return;
+		}
+
+		List<CommerceShippingMethod> commerceShippingMethods =
+			_commerceShippingMethodLocalService.getCommerceShippingMethods(
+				QueryUtil.ALL_POS, QueryUtil.ALL_POS);
+
+		for (CommerceShippingMethod commerceShippingMethod :
+				commerceShippingMethods) {
+
+			if (key.equals(commerceShippingMethod.getEngineKey())) {
+				_commerceShippingMethodLocalService.
+					deleteCommerceShippingMethod(
+						commerceShippingMethod.getCommerceShippingMethodId());
+			}
+		}
+	}
+
+	@Modified
+	protected void modified(Map<String, Object> properties)
+		throws PortalException {
+
+		_functionCommerceShippingEngineConfiguration =
+			ConfigurableUtil.createConfigurable(
+				FunctionCommerceShippingEngineConfiguration.class, properties);
+
+		List<CommerceShippingMethod> commerceShippingMethods =
+			_commerceShippingMethodLocalService.getCommerceShippingMethods(
+				QueryUtil.ALL_POS, QueryUtil.ALL_POS);
+
+		for (CommerceShippingMethod commerceShippingMethod :
+				commerceShippingMethods) {
+
+			String key = (String)properties.get("key");
+
+			if (key.equals(commerceShippingMethod.getEngineKey())) {
+				_commerceShippingMethodLocalService.
+					deleteCommerceShippingMethod(
+						commerceShippingMethod.getCommerceShippingMethodId());
+			}
+		}
+	}
+
+	private List<CommerceShippingOption> _getCommerceShippingOptions(
+		JSONObject jsonObject) {
+
+		List<CommerceShippingOption> commerceShippingOptions =
+			new ArrayList<>();
+
+		JSONArray shippingOptionJSONArray = jsonObject.getJSONArray(
+			"shippingOptions");
+
+		for (Object object : shippingOptionJSONArray) {
+			JSONObject shippingOptionJSONObject = (JSONObject)object;
+
+			commerceShippingOptions.add(
+				new CommerceShippingOption(
+					(BigDecimal)GetterUtil.getNumber(
+						shippingOptionJSONObject.get("amount")),
+					shippingOptionJSONObject.getString("shippingMethodKey"),
+					shippingOptionJSONObject.getString("key"),
+					shippingOptionJSONObject.getString("name"),
+					shippingOptionJSONObject.getDouble("priority")));
```

Comment:
I think you can inline all of this with TransformUtil

Reply:
yes, maybe even better with:
`JSONArray.forEach(
			object -> {
                               .........
})`